checksum.go

```diff
@@ -0,0 +1,233 @@
+package getter
+
+import (
+	"bufio"
+	"crypto/md5"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/hex"
+	"fmt"
+	"hash"
+	"io"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	urlhelper "github.com/hashicorp/go-getter/helper/url"
+)
+
+// fileChecksum helps verifying the checksum for a file.
+type fileChecksum struct {
+	Type     string
+	Hash     hash.Hash
+	Value    []byte
+	Filename string
+}
+
+// extractChecksum will return a fileChecksum based on the 'checksum'
+// parameter of u.
+// ex:
+//  http://hashicorp.com/terraform?checksum=<checksumType>:<checksumValue>
+//  http://hashicorp.com/terraform?checksum=file:<checksum_url>
+// when checksumming from a file, extractChecksum will go get checksum_url
+// in a temporary directory, parse the content of the file then delete it.
+// Content of files are expected to be BSD style or GNU style.
+//
+// BSD-style checksum:
+//  MD5 (file1) = <checksum>
+//  MD5 (file2) = <checksum>
+//
+// GNU-style:
+//  <checksum>  file1
+//  <checksum> *file2
+//
+// see parseChecksumLine for more detail on checksum file parsing
+func extractChecksum(u *url.URL) (*fileChecksum, error) {
+	q := u.Query()
+	v := q.Get("checksum")
+
+	if v == "" {
+		return nil, nil
+	}
+
+	vs := strings.SplitN(v, ":", 2)
+	if len(vs) != 2 {
+		return nil, fmt.Errorf("non explicit checksum type: %s", v)
```

Comment:
here instead, if the user just did `checksum=<checksum>` we could guess checksum based on length ? Would that be a good idea though ?

Reply:
I added a commit that does that, easy to revert; but if you think it's a good idea that would avoid a back and forth 🙂

Reply:
Checking on _byte_ length is fine - from what I understand that this is actually how tools like `shasum` and what not check when looking at a format file. Ultimately, since all algos have a static hash length, if it guesses wrong, the hash will fail anyway, and a false negative is better than a false positive. :slightly_smiling_face: :